File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/booking/ProxyRiderScreen.kt
package lk.mageride.passenger.booking

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.onboarding.PhoneNumber
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.LabelledTextField
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.component.PhoneNumberField
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.ride.LocationRequestState
import org.koin.androidx.compose.koinViewModel

/**
 * SCR-PA-010b — booking for somebody else (US-8.16–8.19, P-01…P-03).
 *
 * The wireframe: `‹ Book for someone else`, the rider's name and mobile with a Contacts affordance,
 * the four **pickup methods**, and — when Request is chosen — *"Waiting for rider to share
 * location… 5:00"* over the CTA.
 *
 * **The unregistered case is a first-class state, not an error** (P-03, US-8.19). A number that
 * belongs to nobody cannot be sent an FCM, so the screen says *"Not a MageRide user — enter pickup
 * manually"* and the Request method removes itself. The booking still works; it just captures the
 * pickup a different way.
 *
 * @param onSearch The Search method — SCR-PA-008 with the result coming back here. The **Map**
 *   method is a modal on this screen: the wireframe assigns no SCR-PA id to a map picker, so there
 *   is no route to navigate to. See [MapPickSheet].
 */
@Composable
@Suppress("LongMethod") // The wireframe's form: name, phone, four methods, countdown, CTA.
internal fun ProxyRiderScreen(
    onBack: () -> Unit,
    onSearch: () -> Unit,
    onDone: () -> Unit,
    model: ProxyRiderViewModel = koinViewModel(),
) {
    val state by model.state.collectAsStateWithLifecycle()
    var pasteOpen by remember { mutableStateOf(false) }
    var mapOpen by remember { mutableStateOf(false) }

    Column(modifier = Modifier.fillMaxSize()) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            IconButton(onClick = onBack) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = stringResource(R.string.action_back),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Text(
                text = stringResource(R.string.proxy_title),
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            LabelledTextField(
                label = stringResource(R.string.proxy_rider_name),
                value = state.riderName,
                onValueChange = model::onNameChanged,
                placeholder = stringResource(R.string.proxy_rider_name_hint),
                keyboardType = KeyboardType.Text,
                imeAction = ImeAction.Next,
            )

            SectionLabel(text = stringResource(R.string.proxy_rider_phone))
            PhoneNumberField(
                value = state.riderPhone,
                onValueChange = model::onPhoneChanged,
                countryCode = PhoneNumber.COUNTRY_CODE,
                placeholder = PhoneNumber.PLACEHOLDER,
            )

            // P-03 / US-8.19. Not an error colour: the booking is fine, one method is not.
            if (state.riderRegistered == false) {
                Text(
                    text = stringResource(R.string.proxy_not_registered),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            SectionLabel(text = stringResource(R.string.proxy_pickup_method))
            LocationMethodRow(
                // The Request method disappears entirely for an unregistered rider rather than
                // being drawn disabled — a control that cannot work is not a control.
                methods = if (state.riderRegistered == false) PACKAGE_PICKUP_METHODS else ALL_METHODS,
                selected = state.method,
                onSelect = { method ->
                    model.setMethod(method)
                    when (method) {
                        PickupMethod.SEARCH -> onSearch()
                        PickupMethod.MAP -> mapOpen = true
                        PickupMethod.PASTE_LINK -> pasteOpen = true
                        PickupMethod.REQUEST -> Unit
                    }
                },
            )

            CapturedPlace(place = state.pickup, emptyLabel = stringResource(R.string.proxy_no_pickup))

            if (state.method == PickupMethod.REQUEST) {
                RequestRow(state = state, onRequest = model::requestRiderLocation)
            }

            state.error?.let { InlineError(message = stringResource(it)) }

            MageRideCta(
                label = stringResource(R.string.proxy_continue),
                onClick = onDone,
                enabled = state.isComplete,
                modifier = Modifier.padding(bottom = MageRideTheme.spacing.md),
            )
        }
    }

    if (pasteOpen) {
        PasteLinkSheet(
            label = stringResource(R.string.paste_label_pickup),
            onUse = { place ->
                model.setPickup(place)
                pasteOpen = false
            },
            onPickOnMap = {
                pasteOpen = false
                mapOpen = true
            },
            onDismiss = { pasteOpen = false },
        )
    }

    if (mapOpen) {
        MapPickSheet(
            label = stringResource(R.string.proxy_pickup_method),
            around = state.pickup?.point,
            onUse = { place ->
                model.setPickup(place)
                mapOpen = false
            },
            onDismiss = { mapOpen = false },
        )
    }
}

/**
 * The P-02 round trip, as the wireframe draws it.
 *
 * Pending is a countdown rather than an indefinite spinner because the window is real and finite:
 * five minutes, enforced by a durable Quartz timer server-side. A booker watching `0:41` knows
 * whether to wait or to switch methods; a booker watching a spinner does not.
 */
@Composable
private fun RequestRow(state: ProxyRiderState, onRequest: () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
        when (state.requestState) {
            LocationRequestState.Pending -> Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
            ) {
                CircularProgressIndicator(modifier = Modifier.size(ControlTokens.ChipIcon))
                Text(
                    text = stringResource(R.string.proxy_waiting, state.countdown),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            LocationRequestState.Confirmed -> Text(
                text = stringResource(R.string.proxy_confirmed),
                style = MaterialTheme.typography.bodyMedium,
                color = MageRideTheme.status.success,
            )

            // Declined and Expired read the same to a booker — the rider did not share — and both
            // land on the same instruction. Naming which one it was would tell the booker their
            // rider refused, which is not theirs to know (P-02).
            LocationRequestState.Declined, LocationRequestState.Expired -> Text(
                text = stringResource(R.string.proxy_no_answer),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            LocationRequestState.RiderNotRegistered, null -> Unit
        }

        if (state.requestState != LocationRequestState.Confirmed) {
            MageRideCta(
                label = stringResource(R.string.proxy_request_location),
                onClick = onRequest,
                enabled = state.canRequestLocation,
            )
        }
    }
}

File: apps/passenger-ios/PassengerApp/Booking/ProxyRiderScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-010b — booking for somebody else (US-8.16–8.19, P-01…P-03).
///
/// The cell: `‹ Back · For someone else`, a grouped list with the rider's name and mobile, the four
/// **pickup methods** as a segmented control, the paste-link note, and — when Request is chosen —
/// *"Waiting for rider… 5:00"* over the CTA.
///
/// **The unregistered case is a first-class state, not an error** (P-03, US-8.19). A number that
/// belongs to nobody cannot be sent an FCM, so the screen says *"Not a MageRide user — enter pickup
/// manually"* and the Request method removes itself. The booking still works; it just captures the
/// pickup a different way.
///
/// **Δ the wireframe's `👥 Pick from Contacts` row is not drawn**, and that is the same call
/// `apps/passenger-android` made: `CNContactPickerViewController` would work here (the driver app
/// uses one on SCR-DI-029), but the Android twin has no picker either and adding one on this side
/// alone is a parity break rather than a port. Recorded in the C097 handoff.
///
/// - Parameter onSearch: The Search method — SCR-PI-008, with the result coming back through the
///   draft. The **Map** method is a sheet on this screen: no SCR-PI id names a map picker, so there
///   is no destination to navigate to. See ``MapPickSheet``.
@MainActor
struct ProxyRiderScreen: View {

    @StateObject private var model: ProxyRiderModel

    private let bookings: BookingRepository

    /// The geocoder behind the Map method's search box, and the fix it opens over. Held rather than
    /// reached for: ``MapPickSheet`` builds its own model, and a picker that opens on
    /// `MapCamera.colombo` opens on Colombo Fort for a booker standing anywhere else — with its CTA
    /// disabled until the first camera settle produces a centre.
    private let places: PassengerPlaces
    private let lastFix: LastKnownFix

    let onBack: () -> Void
    let onSearch: () -> Void
    let onDone: () -> Void

    @State private var isPasteOpen = false
    @State private var isMapOpen = false

    init(
        draft: BookingDraft,
        bookings: BookingRepository,
        live: PassengerLiveMap,
        places: PassengerPlaces,
        lastFix: LastKnownFix,
        onBack: @escaping () -> Void,
        onSearch: @escaping () -> Void,
        onDone: @escaping () -> Void
    ) {
        _model = StateObject(wrappedValue: ProxyRiderModel(draft: draft, bookings: bookings, live: live))
        self.bookings = bookings
        self.places = places
        self.lastFix = lastFix
        self.onBack = onBack
        self.onSearch = onSearch
        self.onDone = onDone
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                LabelledTextField(
                    labelKey: "proxy_rider_name",
                    value: Binding(get: { model.state.riderName }, set: { model.onNameChanged($0) }),
                    placeholder: "proxy_rider_name_hint".localised,
                    textContentType: .name
                )

                SectionLabel(key: "proxy_rider_phone")
                PhoneNumberField(
                    value: Binding(get: { model.state.riderPhone }, set: { model.onPhoneChanged($0) })
                )

                // P-03 / US-8.19. Not an error colour: the booking is fine, one method is not.
                if model.state.riderRegistered == false {
                    Text(key: "proxy_not_registered")
                        .mageFont(.bodySmall)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                }

                SectionLabel(key: "proxy_pickup_method")
                LocationMethodPicker(
                    // The Request method disappears entirely for an unregistered rider rather than
                    // being drawn disabled — a control that cannot work is not a control.
                    methods: model.state.riderRegistered == false
                        ? PickupMethod.packagePickupMethods
                        : PickupMethod.allMethods,
                    selection: model.state.method,
                    onSelect: choose
                )

                Text(key: "capture_paste_note")
                    .mageFont(.caption)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)

                CapturedPlaceRow(place: model.state.pickup, emptyKey: "proxy_no_pickup")

                if model.state.method == .request {
                    requestRow
                }

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                Button(action: onDone) {
                    Text(key: "proxy_continue")
                }
                .buttonStyle(.mageCta)
                .disabled(!model.state.isComplete)
                .padding(.top, MageRideSpacing.xs)
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: "proxy_title"))
        .navigationBarTitleDisplayMode(.inline)
        .task { model.start() }
        .onAppear { model.setPickupFromDraft() }
        .sheet(isPresented: $isPasteOpen) {
            PasteLinkSheet(
                bookings: bookings,
                target: .pickup,
                onUse: { model.setPickup($0) },
                onPickOnMap: { isMapOpen = true },
                onDismiss: { isPasteOpen = false }
            )
        }
        .sheet(isPresented: $isMapOpen) {
            MapPickSheet(
                places: places,
                titleKey: "proxy_pickup_method",
                around: model.state.pickup?.point ?? lastFix.point,
                onUse: { model.setPickup($0) },
                onDismiss: { isMapOpen = false }
            )
        }
    }

    /// The P-02 round trip, as the wireframe draws it.
    ///
    /// Pending is a countdown rather than an indefinite spinner because the window is real and
    /// finite: five minutes, enforced by a durable timer server-side. A booker watching `0:41` knows
    /// whether to wait or to switch methods; a booker watching a spinner does not.
    @ViewBuilder
    private var requestRow: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xs) {
            switch model.state.requestState {
            case LocationRequestState.pending:
                HStack(spacing: MageRideSpacing.xs) {
                    ProgressView()
                    Text("proxy_waiting".localisedFormat(model.state.countdown))
                        .mageFont(.bodySmall)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                        .monospacedDigit()
                }
                .padding(MageRideSpacing.sm)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(
                    MageRideColor.surfaceVariant,
                    in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
                )

            case LocationRequestState.confirmed:
                Text(key: "proxy_confirmed")
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.success)

            // Declined and Expired read the same to a booker — the rider did not share — and both
            // land on the same instruction. Naming which one it was would tell the booker their
            // rider refused, which is not theirs to know (P-02).
            case LocationRequestState.declined, LocationRequestState.expired:
                Text(key: "proxy_no_answer")
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)

            default:
                EmptyView()
            }

            if model.state.requestState != LocationRequestState.confirmed {
                Button { model.requestRiderLocation() } label: {
                    Text(key: "proxy_request_location")
                }
                .buttonStyle(.mageCta)
                .disabled(!model.state.canRequestLocation)
            }
        }
    }

    private func choose(_ method: PickupMethod) {
        model.setMethod(method)
        switch method {
        case .search: onSearch()
        case .map: isMapOpen = true
        case .pasteLink: isPasteOpen = true
        case .request: break
        }
    }
}
